crypto/bp/bp_lcl.h

```diff
@@ -0,0 +1,321 @@
+/********************************************************************/
+
+/*
+ * Opaque type for quadratic extension field element
+ */
+typedef struct bp_fp2_elem_st FP2;
+
+/*
+ * Opaque type for sextic extension field element
+ */
+typedef struct bp_fp6_elem_st FP6;
+
+/*
+ * Opaque type for dodecic extension field element
+ */
+typedef struct bp_fp12_elem_st FP12;
+
+/*
+ * G2 precomputation data
+ */
+typedef struct g2_pre_comp_st G2_PRE_COMP;
+
+/*
+ * Bilinear groups
+ */
+struct bp_group_st {
+    /*
+     * Group of points for curve defined over the base field (G1).
+     */
+    EC_GROUP *ec;
+    /*
+     * Convenient copy of the prime field modulus.
+     */
+    BIGNUM *field;
+    /*
+     * Montgomery data to handle extension field arithmetic.
+     */
+    BN_MONT_CTX *mont;
+    /*
+     * Convenient copy of one converted to Montgomery representation.
+     */
+    BIGNUM *one;
+    /*
+     * Curve parameter.
+     */
+    BIGNUM *param;
+    /*
+     * Generator for G_2.
+     */
+    G2_ELEM *gen2;
+    /*
+     * Precomputed data for G_2.
+     */
+    G2_PRE_COMP *g2_pre_comp;
+    /*
+     * Constants to help with Frobenius map.
+     */
+    FP2 *frb;
+};
+
+/*
+ * G1 element
+ */
+struct bp_g1_elem_st {
+    /*
+     * Point in the curve defined over the base field
+     */
+    EC_POINT *p;
+};
+
+/*
+ * G2 element
+ */
+struct bp_g2_elem_st {
+    /**
+     * X-coordinate of point defined over the twist
+     */
+    FP2 *X;
+    /**
+     * Y-coordinate of point defined over the twist
+     */
+    FP2 *Y;
+    /**
+     * Z-coordinate of point defined over the twist
+     */
+    FP2 *Z;
+    /**
+     * Flag to indicate if point in affine coordinates
+     */
+    int Z_is_one;
+};
+
+/**
+ * Quadratic extension field element.
+ */
+struct bp_fp2_elem_st {
+    /**
+     * Vector of multiprecision integers.
+     */
+    BIGNUM *f[2];
+};
+
+/**
+ * Sextic extension field element.
+ */
+struct bp_fp6_elem_st {
+    /**
+     * Vector of quadratic extension field elements.
+     */
+    FP2 *f[3];
+};
+
+/**
+ * Dodecic extension field element.
+ */
+struct bp_fp12_elem_st {
+    /**
+     * Vector of sextic extension field elements.
+     */
+    FP6 *f[2];
+};
+
+/**
+ * GT element.
+ */
+struct bp_gt_elem_st {
+    /**
+     * Underlying dodecic extension field element.
+     */
+    FP12 *f;
+};
+
+/********************************************************************/
+/*               Functions for managing Fp^2 elements               */
+/********************************************************************/
+
+FP2 *FP2_new(void);
+void FP2_init(FP2 *a);
+void FP2_clear(FP2 *a);
+void FP2_free(FP2 *a);
+void FP2_clear_free(FP2 *a);
+
+/********************************************************************/
+/*               Functions for arithmetic in Fp^2                   */
+/********************************************************************/
+
+int FP2_rand(const BP_GROUP *group, FP2 *a);
```

Comment:
In internal (but not public) headers, consider using `__owur` (OpenSSL-Warn-Unused-Result) where appropriate.